-- Inferring locals/globals with simple types
-- ------------------------------------------


[case testInferSimpleGvarType]
import typing
x = A()
y = B()
x = B() # Fail
x = A()
x = y   # Fail
x = x
class A: pass
class B: pass
[out]
main, line 4: Incompatible types in assignment (expression has type "B", variable has type "A")
main, line 6: Incompatible types in assignment (expression has type "B", variable has type "A")

[case testInferSimpleLvarType]
import typing
def f() -> None:
  x = A()
  y = B()
  x = B() # Fail
  x = A()
  x = y   # Fail
  x = x
class A: pass
class B: pass
[out]
main: In function "f":
main, line 5: Incompatible types in assignment (expression has type "B", variable has type "A")
main, line 7: Incompatible types in assignment (expression has type "B", variable has type "A")

[case testLvarInitializedToNoneWithoutType]
import typing
def f() -> None:
    a = None # E: Need type annotation for variable
    a.x()
[out]
main: In function "f":

[case testLvarInitializedToVoid]
import typing
def f() -> None:
    a = g()    # E: "g" does not return a value
    #b, c = g() # "g" does not return a value TODO

def g() -> None: pass
[out]
main: In function "f":

[case testInferringLvarTypeFromArgument]
import typing
def f(a: 'A') -> None:
    b = a
    b = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")
    b = a
    a = b

class A: pass
class B: pass
[out]
main: In function "f":

[case testInferringLvarTypeFromGvar]
from typing import Undefined
g = Undefined # type: B

def f() -> None:
    a = g
    a = A() # E: Incompatible types in assignment (expression has type "A", variable has type "B")
    a = B()

class A: pass
class B: pass
[out]
main: In function "f":

[case testInferringImplicitDynamicTypeForLvar]
import typing
def f() -> None:
    a = g()
    None(a) # E: None not callable
    a.x()

def g(): pass
[out]
main: In function "f":

[case testInferringExplicitDynamicTypeForLvar]
from typing import Undefined, Any
g = Undefined # type: Any

def f(a: Any) -> None:
    b = g
    None(b) # E: None not callable
    a.x()
[out]
main: In function "f":


-- Inferring types of local variables with complex types
-- -----------------------------------------------------


[case testInferringTupleTypeForLvar]
from typing import Undefined
def f() -> None:
    a = A(), B()
    aa = Undefined # type: A
    bb = Undefined # type: B
    bb = a[0] # E: Incompatible types in assignment (expression has type "A", variable has type "B")
    aa = a[1] # E: Incompatible types in assignment (expression has type "B", variable has type "A")
    aa = a[0]
    bb = a[1]

class A: pass
class B: pass
[builtins fixtures/tuple.py]
[out]
main: In function "f":

[case testInferringTupleTypeForLvarWithNones]
import typing
def f() -> None:
    a = A(), None # E: Need type annotation for variable
    b = None, A() # E: Need type annotation for variable

class A: pass
[builtins fixtures/tuple.py]
[out]
main: In function "f":

[case testInferringGenericTypeForLvar]
from typing import Undefined, typevar, Generic
T = typevar('T')
class A(Generic[T]): pass
a_i = Undefined # type: A[int]
a_s = Undefined # type: A[str]

def f() -> None:
    a = A[int]()
    a = a_s # E: Incompatible types in assignment (expression has type A[str], variable has type A[int])
    a = a_i
[builtins fixtures/tuple.py]
[out]
main: In function "f":

[case testInferringFunctionTypeForLvar]
import typing
def f() -> None:
    a = g
    a(B()) # E: Argument 1 has incompatible type "B"; expected "A"
    a(A())

def g(a: 'A') -> None: pass

class A: pass
class B: pass
[out]
main: In function "f":

[case testInferringFunctionTypeForLvarFromTypeObject]
import typing
def f() -> None:
    a = A
    a(A()) # E: Too many arguments
    a()
    t = a # type: type

class A: pass
[out]
main: In function "f":


-- Inferring variable types in multiple definition
-- -----------------------------------------------


[case testInferringLvarTypesInMultiDef]
import typing
def f() -> None:
    a, b = A(), B()
    a = b   # E: Incompatible types in assignment (expression has type "B", variable has type "A")
    a = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")
    b = A() # E: Incompatible types in assignment (expression has type "A", variable has type "B")

    a = A()
    b = B()

class A: pass
class B: pass
[out]
main: In function "f":

[case testInferringLvarTypesInTupleAssignment]
from typing import Undefined, Tuple
def f() -> None:
    t = Undefined # type: Tuple[A, B]
    a, b = t
    a = b   # E: Incompatible types in assignment (expression has type "B", variable has type "A")
    a = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")
    b = A() # E: Incompatible types in assignment (expression has type "A", variable has type "B")

    a = A()
    b = B()

class A: pass
class B: pass
[out]
main: In function "f":

[case testInferringLvarTypesInNestedTupleAssignment]
from typing import Undefined, Tuple
def f() -> None:
    t = Undefined # type: Tuple[A, B]
    a1, (a, b) = A(), t
    a = b   # E: Incompatible types in assignment (expression has type "B", variable has type "A")
    a = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")
    b = A() # E: Incompatible types in assignment (expression has type "A", variable has type "B")

    a = A()
    b = B()

class A: pass
class B: pass
[out]
main: In function "f":

[case testInferringLvarTypesInNestedTupleAssignment]
import typing
def f() -> None:
    a, (b, c) = A(), (B(), C())
    a = b   # E: Incompatible types in assignment (expression has type "B", variable has type "A")
    a = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")
    b = A() # E: Incompatible types in assignment (expression has type "A", variable has type "B")
    c = A() # E: Incompatible types in assignment (expression has type "A", variable has type "C")

    a = A()
    b = B()
    c = C()

class A: pass
class B: pass
class C: pass
[out]
main: In function "f":

[case testInferringLvarTypesInMultiDefWithNoneTypes]
import typing
def f() -> None:
    a, b = A(), None # E: Need type annotation for variable
    c, d = None, A() # E: Need type annotation for variable

class A: pass
[out]
main: In function "f":

[case testInferringLvarTypesInNestedTupleAssignmentWithNoneTypes]
import typing
def f() -> None:
    a1, (a2, b) = A(), (A(), None) # E: Need type annotation for variable

class A: pass
[out]
main: In function "f":

[case testInferringLvarTypesInMultiDefWithInvalidTuple]
from typing import Undefined, Tuple
t = Undefined # type: Tuple[object, object, object]

def f() -> None:
    a, b = t         # Fail
    c, d, e, f = t   # Fail
    g, h, i = t
[builtins fixtures/tuple.py]
[out]
main: In function "f":
main, line 5: Too many values to unpack (2 expected, 3 provided)
main, line 6: Need more than 3 values to unpack (4 expected)

[case testInvalidRvalueTypeInInferredMultipleLvarDefinition]
import typing
def f() -> None:
    a, b = f   # E: 'def ()' object is not iterable
    c, d = A() # E: '__main__.A' object is not iterable
class A: pass
[builtins fixtures/for.py]
[out]
main: In function "f":

[case testInvalidRvalueTypeInInferredNestedTupleAssignment]
import typing
def f() -> None:
    a1, (a2, b) = A(), f   # E: 'def ()' object is not iterable
    a3, (c, d) = A(), A() # E: '__main__.A' object is not iterable
class A: pass
[builtins fixtures/for.py]
[out]
main: In function "f":

[case testInferringMultipleLvarDefinitionWithListRvalue]
from typing import List

class C: pass
class D: pass

def f() -> None:
    a, b = List[C]()
    c, d, e = List[D]()
    a = D() # E: Incompatible types in assignment (expression has type "D", variable has type "C")
    b = D() # E: Incompatible types in assignment (expression has type "D", variable has type "C")
    c = C() # E: Incompatible types in assignment (expression has type "C", variable has type "D")
    b = c   # E: Incompatible types in assignment (expression has type "D", variable has type "C")

    a = C()
    b = C()
    c = D()
    d = D()
    e = D()

    a = b
    c = d
    d = e
[builtins fixtures/for.py]
[out]
main: In function "f":

[case testInferringNestedTupleAssignmentWithListRvalue]
from typing import List

class C: pass
class D: pass

def f() -> None:
    c1, (a, b) = C(), List[C]()
    c2, (c, d, e) = C(), List[D]()
    a = D() # E: Incompatible types in assignment (expression has type "D", variable has type "C")
    b = D() # E: Incompatible types in assignment (expression has type "D", variable has type "C")
    c = C() # E: Incompatible types in assignment (expression has type "C", variable has type "D")
    b = c   # E: Incompatible types in assignment (expression has type "D", variable has type "C")

    a = C()
    b = C()
    c = D()
    d = D()
    e = D()

    a = b
    c = d
    d = e
[builtins fixtures/for.py]
[out]
main: In function "f":

[case testInferringMultipleLvarDefinitionWithImplicitDynamicRvalue]
import typing
def f() -> None:
    a, b = g()
    a.x
    b.x
def g(): pass

[case testInferringMultipleLvarDefinitionWithExplicitDynamicRvalue]
from typing import Any
def f(d: Any) -> None:
    a, b = d
    a.x
    b.x

[case testInferringTypesFromIterable]
from typing import Iterable
class Nums(Iterable[int]):
    def __iter__(self): pass
    def __next__(self): pass
a, b = Nums()
a = b = 1
a = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
b = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
[builtins fixtures/for.py]


-- Type variable inference for generic functions
-- ---------------------------------------------


[case testInferSimpleGenericFunction]
from typing import Undefined, Tuple, typevar
T = typevar('T')
a = Undefined # type: A
b = Undefined # type: B
c = Undefined # type: Tuple[A, object]

b = id(a) # E: Argument 1 to "id" has incompatible type "A"; expected "B"
a = id(b) # E: Argument 1 to "id" has incompatible type "B"; expected "A"
a = id(c) # E: Argument 1 to "id" has incompatible type "Tuple[A, object]"; expected "A"

a = id(a)
b = id(b)
c = id(c)

def id(a: T) -> T: pass

class A: pass
class B: pass
[builtins fixtures/tuple.py]

[case testInferringGenericFunctionTypeForLvar]
# TODO this is actually a contextual inference test case
from typing import Undefined, typevar
T = typevar('T')
def f() -> None:
    a = id
    b = Undefined # type: int
    c = Undefined # type: str
    b = a(c) # E: Argument 1 has incompatible type "str"; expected "int"
    b = a(b)
    c = a(c)
def id(x: T) -> T:
    return x
[out]
main: In function "f":

[case testUnderspecifiedInferenceResult]
from typing import Undefined, typevar
T = typevar('T')
class A: pass
a = Undefined # type: A

def ff() -> None:
    x = f() # E: Need type annotation for variable

g(None) # Ok
f()     # Ok because not used to infer local variable type
f[A]()
g(a)

def f() -> T: pass
def g(a: T) -> None: pass
[out]
main: In function "ff":

[case testUnsolvableInferenceResult]
from typing import typevar
T = typevar('T')
f(A(), g()) # Fail
f(A(), A())

def f(a: T, b: T) -> None: pass
def g() -> None: pass
class A: pass
[out]
main, line 3: Cannot infer type argument 1 of "f"
main, line 3: "g" does not return a value

[case testInferenceWithMultipleConstraints]
from typing import Undefined, typevar
T = typevar('T')
a = Undefined # type: A
b = Undefined # type: B

b = f(a, b) # E: Argument 1 to "f" has incompatible type "A"; expected "B"
b = f(b, a) # E: Argument 2 to "f" has incompatible type "A"; expected "B"
a = f(a, b)
a = f(b, a)

def f(a: T, b: T) -> T: pass

class A: pass
class B(A): pass

[case testInferenceWithMultipleVariables]
from typing import Undefined, Tuple, typevar
T = typevar('T')
S = typevar('S')
a, b = Undefined, Undefined # type: (A, B)
taa = Undefined # type: Tuple[A, A]
tab = Undefined # type: Tuple[A, B]
tba = Undefined # type: Tuple[B, A]

taa = f(a, b) # Fail
taa = f(b, a) # Fail
tba = f(a, b) # Fail

tab = f(a, b)
tba = f(b, a)

def f(a: T, b: S) -> Tuple[T, S]: pass

class A: pass
class B: pass
[builtins fixtures/tuple.py]
[out]
main, line 9: Argument 2 to "f" has incompatible type "B"; expected "A"
main, line 10: Argument 1 to "f" has incompatible type "B"; expected "A"
main, line 11: Argument 1 to "f" has incompatible type "A"; expected "B"
main, line 11: Argument 2 to "f" has incompatible type "B"; expected "A"

[case testConstraintSolvingWithSimpleGenerics]
from typing import Undefined, typevar, Generic
T = typevar('T')
ao = Undefined # type: A[object]
ab = Undefined # type: A[B]
ac = Undefined # type: A[C]

ab = f(ao) # E: Argument 1 to "f" has incompatible type A[object]; expected A[B]
ao = f(ab) # E: Argument 1 to "f" has incompatible type A[B]; expected A[object]
ab = f(ac) # E: Argument 1 to "f" has incompatible type A[C]; expected A[B]
ab = g(ao) # E: Argument 1 to "g" has incompatible type A[object]; expected A[B]
ao = g(ab) # E: Argument 1 to "g" has incompatible type A[B]; expected A[object]

ab = f(ab)
ac = f(ac)
ao = f(ao)

ab = g(ab)
ao = g(ao)

def f(a: 'A[T]') -> 'A[T]': pass

def g(a: T) -> T: pass

class A(Generic[T]): pass
class B: pass
class C: pass

[case testConstraintSolvingFailureWithSimpleGenerics]
from typing import Undefined, typevar, Generic
T = typevar('T')
ao = Undefined # type: A[object]
ab = Undefined # type: A[B]

f(ao, ab) # E: Cannot infer type argument 1 of "f"
f(ab, ao) # E: Cannot infer type argument 1 of "f"
f(ao, ao)
f(ab, ab)

def f(a: 'A[T]', b: 'A[T]') -> None: pass

class A(Generic[T]): pass
class B: pass

[case testTypeInferenceWithCalleeDefaultArgs]
from typing import Undefined, typevar
T = typevar('T')
a = Undefined # type: A
o = Undefined # type: object

a = f(o)    # E: Argument 1 to "f" has incompatible type "object"; expected "A"
a = g(a, o) # E: Argument 2 to "g" has incompatible type "object"; expected "A"

o = f()
o = f(o)
a = f(a)
a = g(a)

def f(a: T = None) -> T: pass
def g(a: T, b: T = None) -> T: pass

class A: pass


-- Generic function inference with multiple inheritance
-- ----------------------------------------------------


[case testGenericFunctionInferenceWithMultipleInheritance]
from typing import typevar

class I: pass
class J: pass

class A(I, J): pass
class B(I, J): pass
class C(I): pass
class D(J): pass

T = typevar('T')
def f(a: T, b: T) -> T: pass
def g(x: I) -> None: pass

a = f(A(), C())
g(a)
b = f(A(), B())
g(b)
c = f(A(), D())
g(c) # E: Argument 1 to "g" has incompatible type "object"; expected "I"
d = f(D(), A())
g(d) # E: Argument 1 to "g" has incompatible type "object"; expected "I"
e = f(D(), C())
g(e) # E: Argument 1 to "g" has incompatible type "object"; expected "I"

[case testGenericFunctionInferenceWithMultipleInheritance2]
from typing import typevar

class I: pass
class J: pass

class A(I): pass
class B(A, J): pass
class C(I, J): pass

T = typevar('T')
def f(a: T, b: T) -> T: pass
def g(x: I) -> None: pass
def h(x: J) -> None: pass

a = f(B(), C())
g(a)
h(a) # E: Argument 1 to "h" has incompatible type "I"; expected "J"
b = f(C(), B())
g(b)
h(b) # E: Argument 1 to "h" has incompatible type "I"; expected "J"
c = f(A(), B())
g(a)
h(b) # E: Argument 1 to "h" has incompatible type "I"; expected "J"

[case testGenericFunctionInferenceWithMultipleInheritance3]
from typing import typevar

class I: pass
class J: pass
class K(J): pass

class A(K): pass
class B(A, I): pass
class C(I, J): pass

T = typevar('T')
def f(a: T, b: T) -> T: pass
def g(x: K) -> None: pass

a = f(B(), C())
g(a) # E: Argument 1 to "g" has incompatible type "object"; expected "K"
b = f(A(), C())
g(b) # E: Argument 1 to "g" has incompatible type "object"; expected "K"
c = f(A(), B())
g(c)

[case testPrecedenceOfFirstBaseAsInferenceResult]
from typing import Undefined, typevar
from abc import abstractmethod, ABCMeta
T = typevar('T')
a, i, j = Undefined, Undefined, Undefined # type: (A, I, J)

a = f(B(), C())

class I(metaclass=ABCMeta): pass
class J(metaclass=ABCMeta): pass

def f(a: T, b: T) -> T: pass

class A: pass
class B(A, I, J): pass
class C(A, I, J): pass


-- Generic function inference with function arguments
-- --------------------------------------------------


[case testNonOverloadedMapInference]
from typing import typevar, Function, List
t = typevar('t')
s = typevar('s')
class A: pass
b = bool()
def f(x: bool) -> A: pass
def mymap(f: Function[[t], s], a: List[t]) -> List[s]: pass

l = mymap(f, [b])
l = List[A]()
l = List[bool]() # E: Incompatible types in assignment (expression has type List[bool], variable has type List[A])
[builtins fixtures/for.py]


-- Literal expressions
-- -------------------


[case testDictLiteral]
from typing import Undefined, Dict
class A: pass
class B: pass
a, b = Undefined, Undefined # type: (A, B)
d = {a:b}
d = Dict[A, B]()
d = Dict[A, A]() # E: Incompatible types in assignment (expression has type Dict[A, A], variable has type Dict[A, B])
d = Dict[B, B]() # E: Incompatible types in assignment (expression has type Dict[B, B], variable has type Dict[A, B])
[builtins fixtures/dict.py]

[case testSetLiteral]
from typing import Undefined, Any, Set
a, x = Undefined, Undefined # type: (int, Any)
s = {a}
s = {x}
s = Set[int]()
s = Set[str]() # E: Incompatible types in assignment (expression has type Set[str], variable has type Set[int])
[builtins fixtures/set.py]


-- For statements
-- --------------


[case testInferenceOfFor1]
from typing import Undefined
a, b = Undefined, Undefined # type: (A, B)

for x in [A()]:
    b = x # E: Incompatible types in assignment (expression has type "A", variable has type "B")
    a = x

for y in []: # E: Need type annotation for variable
    a = y # E: Cannot determine type of 'y'

class A: pass
class B: pass
[builtins fixtures/for.py]

[case testInferenceOfFor2]
from typing import Undefined
a, b = Undefined, Undefined # type: (A, B)
for x, y in [(A(), B())]:
    b = x # Fail
    a = y # Fail
    a = x
    b = y
for xx, yy, zz in [(A(), B())]: # Fail
    pass
for xxx, yyy in [(None, None)]: # Fail
    pass

class A: pass
class B: pass
[builtins fixtures/for.py]
[out]
main, line 4: Incompatible types in assignment (expression has type "A", variable has type "B")
main, line 5: Incompatible types in assignment (expression has type "B", variable has type "A")
main, line 8: Need more than 2 values to unpack (3 expected)
main, line 10: Need type annotation for variable

[case testInferenceOfFor3]
from typing import Undefined
a, b = Undefined, Undefined # type: (A, B)

for x, y in [[A()]]:
    b = x # E: Incompatible types in assignment (expression has type "A", variable has type "B")
    b = y # E: Incompatible types in assignment (expression has type "A", variable has type "B")
    a = x
    a = y

for e, f in [[]]: # E: Need type annotation for variable
    pass

class A: pass
class B: pass
[builtins fixtures/for.py]

[case testForStatementInferenceWithVoid]
import typing
for x in f(): # E: "f" does not return a value
    pass
def f() -> None: pass
[builtins fixtures/for.py]

[case testReusingInferredForIndex]
import typing
for a in [A()]: pass
a = A()
a = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")
for a in []: pass
a = A()
a = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")
class A: pass
class B: pass
[builtins fixtures/for.py]

[case testReusingInferredForIndex2]
import typing
def f() -> None:
    for a in [A()]: pass
    a = A()
    a = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")
    for a in []: pass
    a = A()
    a = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")
class A: pass
class B: pass
[builtins fixtures/for.py]
[out]
main: In function "f":


-- Regression tests
-- ----------------


[case testMultipleAssignmentWithPartialDefinition]
from typing import Undefined
a = Undefined # type: A
x, a = a, a
x = a
a = x
x = object() # E: Incompatible types in assignment (expression has type "object", variable has type "A")
a = object() # E: Incompatible types in assignment (expression has type "object", variable has type "A")
class A: pass

[case testMultipleAssignmentWithPartialDefinition2]
from typing import Undefined
a = Undefined # type: A
a, x = [a, a]
x = a
a = x
x = object() # E: Incompatible types in assignment (expression has type "object", variable has type "A")
a = object() # E: Incompatible types in assignment (expression has type "object", variable has type "A")
class A: pass
[builtins fixtures/for.py]

[case testMultipleAssignmentWithPartialDefinition3]
from typing import Undefined, Any
a = Undefined # type: A
x, a = Any(a)
x = a
a = x
x = object()
a = object() # E: Incompatible types in assignment (expression has type "object", variable has type "A")
class A: pass

[case testInferGlobalDefinedInBlock]
import typing
if A:
    a = A()
    a = A()
    a = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")
class A: pass
class B: pass


-- Inferring attribute types
-- -------------------------


[case testInferAttributeType]
import typing
class A:
    a = B()
class B: pass

A().a = B()
A().a = A() # E: Incompatible types in assignment (expression has type "A", variable has type "B")

[case testInferAttributeTypeAndAssignInInit]
import typing
class A:
    a = B()
    def __init__(self) -> None:
        self.a = A() # E: Incompatible types in assignment (expression has type "A", variable has type "B")
        self.a = B()
class B: pass
[out]
main: In member "__init__" of class "A":

[case testInferAttributeInInit]
import typing
class B: pass
class A:
    def __init__(self) -> None:
        self.a = A()
        self.b = B()
a = A()
a.a = A()
a.b = B()
a.a = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")
a.b = A() # E: Incompatible types in assignment (expression has type "A", variable has type "B")

[case testInferAttributeInInitUsingChainedAssignment]
import typing
class B: pass
class A:
    def __init__(self) -> None:
        self.a = self.b = A()
a = A()
a.a = A()
a.b = A()
a.a = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")
a.b = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")


-- Lambdas
-- -------


[case testInferLambdaType]
from typing import List, Function, Undefined
l = lambda: List[int]()
f1 = l # type: Function[[], List[int]]
f2 = l # type: Function[[], List[str]] # E: Incompatible types in assignment (expression has type , variable has type )
[builtins fixtures/list.py]

[case testInferLambdaType2]
from typing import List, Function
l = lambda: [B()]
f1 = l # type: Function[[], List[B]]
f2 = l # type: Function[[], List[A]] # E: Incompatible types in assignment (expression has type , variable has type )

class A: pass
class B: pass
[builtins fixtures/list.py]


-- Boolean operators
-- -----------------


[case testOrOperationWithGenericOperands]
from typing import Undefined, List
a = Undefined # type: List[A]
o = Undefined # type: List[object]
a2 = a or []
a = a2
a2 = o # E: Incompatible types in assignment (expression has type List[object], variable has type List[A])
class A: pass
[builtins fixtures/list.py]


-- Accessing variable before its type has been inferred
-- ----------------------------------------------------


[case testAccessGlobalVarBeforeItsTypeIsAvailable]
import typing
x.y  # E: Cannot determine type of 'x'
x = object()
x.y  # E: "object" has no attribute "y"

[case testAccessDataAttributeBeforeItsTypeIsAvailable]
from typing import Undefined
a = Undefined # type: A
a.x.y  # E: Cannot determine type of 'x'
class A:
   def __init__(self) -> None:
       self.x = object()
a.x.y  # E: "object" has no attribute "y"


-- Ducktype declarations
-- ---------------------


[case testListWithDucktypeCompatibility]
from typing import List, Undefined, ducktype
class A: pass
@ducktype(A)
class B: pass
a = Undefined(List[A])
x1 = [A(), B()]
x2 = [B(), A()]
x3 = [B(), B()]
a = x1
a = x2
a = x3 # E: Incompatible types in assignment (expression has type List[B], variable has type List[A])
[builtins fixtures/list.py]

[case testListWithDucktypeCompatibilityAndTransitivity]
from typing import List, Undefined, ducktype
class A: pass
@ducktype(A)
class B: pass
@ducktype(B)
class C: pass
a = Undefined(List[A])
x1 = [A(), C()]
x2 = [C(), A()]
x3 = [B(), C()]
a = x1
a = x2
a = x3 # E: Incompatible types in assignment (expression has type List[B], variable has type List[A])
[builtins fixtures/list.py]
