From 12fb8e77eb5efc5c025d961e04ba7ccef23333bf Mon Sep 17 00:00:00 2001
From: Shantanu Jain <hauntsaninja@gmail.com>
Date: Sun, 22 Feb 2026 20:37:48 -0800
Subject: [PATCH] Temporarily revert contextlib deprecation

---
 mypy/typeshed/stdlib/contextlib.pyi | 15 +--------------
 1 file changed, 1 insertion(+), 14 deletions(-)

diff --git a/mypy/typeshed/stdlib/contextlib.pyi b/mypy/typeshed/stdlib/contextlib.pyi
index b95c23502..73cdda3b8 100644
--- a/mypy/typeshed/stdlib/contextlib.pyi
+++ b/mypy/typeshed/stdlib/contextlib.pyi
@@ -5,7 +5,7 @@ from abc import ABC, abstractmethod
 from collections.abc import AsyncGenerator, AsyncIterator, Awaitable, Callable, Generator, Iterator
 from types import TracebackType
 from typing import Any, Generic, ParamSpec, Protocol, TypeAlias, TypeVar, overload, runtime_checkable, type_check_only
-from typing_extensions import Self, deprecated
+from typing_extensions import Self
 
 __all__ = [
     "aclosing",
@@ -84,12 +84,6 @@ class _GeneratorContextManager(
         self, typ: type[BaseException] | None, value: BaseException | None, traceback: TracebackType | None
     ) -> bool | None: ...
 
-@overload
-def contextmanager(func: Callable[_P, Generator[_T_co, None, object]]) -> Callable[_P, _GeneratorContextManager[_T_co]]: ...
-@overload
-@deprecated(
-    "Annotating the return type as `-> Iterator[Foo]` with `@contextmanager` is deprecated. Use `-> Generator[Foo]` instead."
-)
 def contextmanager(func: Callable[_P, Iterator[_T_co]]) -> Callable[_P, _GeneratorContextManager[_T_co]]: ...
 
 _AF = TypeVar("_AF", bound=Callable[..., Awaitable[Any]])
@@ -107,13 +101,6 @@ class _AsyncGeneratorContextManager(
         self, typ: type[BaseException] | None, value: BaseException | None, traceback: TracebackType | None
     ) -> bool | None: ...
 
-@overload
-def asynccontextmanager(func: Callable[_P, AsyncGenerator[_T_co]]) -> Callable[_P, _AsyncGeneratorContextManager[_T_co]]: ...
-@overload
-@deprecated(
-    "Annotating the return type as `-> AsyncIterator[Foo]` with `@asynccontextmanager` is deprecated. "
-    "Use `-> AsyncGenerator[Foo]` instead."
-)
 def asynccontextmanager(func: Callable[_P, AsyncIterator[_T_co]]) -> Callable[_P, _AsyncGeneratorContextManager[_T_co]]: ...
 
 @type_check_only
-- 
2.54.0

