From ae45edc46f2ca9c939e317648b44ed3372487798 Mon Sep 17 00:00:00 2001
From: Shantanu Jain <hauntsaninja@gmail.com>
Date: Sun, 22 Feb 2026 20:37:48 -0800
Subject: [PATCH] Temporarily revert contextlib deprecation

---
 mypy/typeshed/stdlib/contextlib.pyi | 15 +--------------
 1 file changed, 1 insertion(+), 14 deletions(-)

diff --git a/mypy/typeshed/stdlib/contextlib.pyi b/mypy/typeshed/stdlib/contextlib.pyi
index 8d5902c46..221102ee2 100644
--- a/mypy/typeshed/stdlib/contextlib.pyi
+++ b/mypy/typeshed/stdlib/contextlib.pyi
@@ -5,7 +5,7 @@ from abc import ABC, abstractmethod
 from collections.abc import AsyncGenerator, AsyncIterator, Awaitable, Callable, Generator, Iterator
 from types import TracebackType
 from typing import Any, Generic, Protocol, TypeVar, overload, runtime_checkable, type_check_only
-from typing_extensions import ParamSpec, Self, TypeAlias, deprecated
+from typing_extensions import ParamSpec, Self, TypeAlias
 
 __all__ = [
     "contextmanager",
@@ -86,12 +86,6 @@ class _GeneratorContextManager(
         self, typ: type[BaseException] | None, value: BaseException | None, traceback: TracebackType | None
     ) -> bool | None: ...
 
-@overload
-def contextmanager(func: Callable[_P, Generator[_T_co, None, object]]) -> Callable[_P, _GeneratorContextManager[_T_co]]: ...
-@overload
-@deprecated(
-    "Annotating the return type as `-> Iterator[Foo]` with `@contextmanager` is deprecated. Use `-> Generator[Foo]` instead."
-)
 def contextmanager(func: Callable[_P, Iterator[_T_co]]) -> Callable[_P, _GeneratorContextManager[_T_co]]: ...
 
 if sys.version_info >= (3, 10):
@@ -118,13 +112,6 @@ else:
             self, typ: type[BaseException] | None, value: BaseException | None, traceback: TracebackType | None
         ) -> bool | None: ...
 
-@overload
-def asynccontextmanager(func: Callable[_P, AsyncGenerator[_T_co]]) -> Callable[_P, _AsyncGeneratorContextManager[_T_co]]: ...
-@overload
-@deprecated(
-    "Annotating the return type as `-> AsyncIterator[Foo]` with `@asynccontextmanager` is deprecated. "
-    "Use `-> AsyncGenerator[Foo]` instead."
-)
 def asynccontextmanager(func: Callable[_P, AsyncIterator[_T_co]]) -> Callable[_P, _AsyncGeneratorContextManager[_T_co]]: ...
 @type_check_only
 class _SupportsClose(Protocol):
-- 
2.53.0

