From 274a77fe74d961e9885036ed31e5ed46f8d91daa Mon Sep 17 00:00:00 2001
From: Marc Mueller <30130371+cdce8p@users.noreply.github.com>
Date: Mon, 23 Feb 2026 12:00:18 +0100
Subject: [PATCH] Revert Use contravariant type variable in Container

---
 mypy/typeshed/stdlib/typing.pyi | 9 +++------
 1 file changed, 3 insertions(+), 6 deletions(-)

diff --git a/mypy/typeshed/stdlib/typing.pyi b/mypy/typeshed/stdlib/typing.pyi
index 83e114aac..e9f67877c 100644
--- a/mypy/typeshed/stdlib/typing.pyi
+++ b/mypy/typeshed/stdlib/typing.pyi
@@ -640,17 +640,14 @@ class AsyncGenerator(AsyncIterator[_YieldT_co], Protocol[_YieldT_co, _SendT_cont
     ) -> Coroutine[Any, Any, _YieldT_co]: ...
     def aclose(self) -> Coroutine[Any, Any, None]: ...
 
-_ContainerT_contra = TypeVar("_ContainerT_contra", contravariant=True, default=Any)
-
 @runtime_checkable
-class Container(Protocol[_ContainerT_contra]):
+class Container(Protocol[_T_co]):
     # This is generic more on vibes than anything else
     @abstractmethod
-    def __contains__(self, x: _ContainerT_contra, /) -> bool: ...
+    def __contains__(self, x: object, /) -> bool: ...
 
 @runtime_checkable
-class Collection(Iterable[_T_co], Container[Any], Protocol[_T_co]):
-    # Note: need to use Container[Any] instead of Container[_T_co] to ensure covariance.
+class Collection(Iterable[_T_co], Container[_T_co], Protocol[_T_co]):
     # Implement Sized (but don't have it as a base class).
     @abstractmethod
     def __len__(self) -> int: ...
-- 
2.53.0

